python/ray/tune/context.py

```diff
@@ -0,0 +1,113 @@
+import threading
+from typing import Any, Dict, Optional
+
+from ray.train._internal import session
+from ray.train.constants import _v2_migration_warnings_enabled
+from ray.train.context import TrainContext as TrainV1Context
+from ray.train.utils import _copy_doc
+from ray.tune.execution.placement_groups import PlacementGroupFactory
+from ray.util.annotations import Deprecated, PublicAPI
+
+# The context singleton on this process.
+_tune_context: Optional["TuneContext"] = None
+_tune_context_lock = threading.Lock()
+
+
+_TRAIN_SPECIFIC_CONTEXT_DEPRECATION_MESSAGE = (
+    "`{}` is deprecated for Ray Tune because there is no concept of worker ranks "
+    "for Ray Tune, so these methods only make sense to use in the context of "
+    "a Ray Train worker."
+)
+
+
+@PublicAPI(stability="stable")
+class TuneContext(TrainV1Context):
+    """Context to access metadata within Ray Tune functions."""
+
+    # NOTE: These methods are deprecated on the TrainContext, but are still
+    # available on the TuneContext. Re-defining them here to avoid the
+    # deprecation warnings.
```

Comment:
Do you think it would be beneficial to have a common interface/base class that both `TuneContext` and `TrainContext` extend instead of having `TuneContext` directly subclass `TrainContext`? The intention being to clearly differentiate what logic exists for Tune, or Train, or both.

Reply:
Might make sense for `TuneContext` and `TrainV2Context`, but maybe not for `TrainV1Context` since it still needs the backwards compatibility since it's still using Tune.

```
# abstract class
class BaseContext:
    def get_experiment_name():
        raise NotImplementedError

    def get_storage():
        raise NotImplementedError

class TuneContext(BaseContext):
    def get_trial_name(): ...

class TrainV2Context(BaseContext):
    def 
```